test262/per-file/main.js

```diff
@@ -179,37 +181,136 @@ function generateChildren(node) {
+        childName,
+        child,
+        `${pathInTree.join("/")}/${childName}`
       );
-      childNode.querySelector(
-        ".tree-node-github-url"
-      ).href = `https://github.com/tc39/test262/tree/main/${pathInTree.join(
-        "/"
-      )}/${childName}`;
+      node.appendChild(childNode);
 
+      const isLeaf = child.children === null;
       if (!isLeaf) {
-        const actionNode = childNode.querySelector(".tree-node-action");
-        actionNode.href = generateQueryString([...pathInTree, childName]);
-        actionNode.onclick = function (event) {
-          if (event.metaKey || event.ctrlKey) return;
-          event.preventDefault();
-          pathInTree.push(childName);
-          navigate();
-          generateChildren(node);
-        };
+        makeChildNavigable(childNode, [childName], node);
+      }
+    });
+}
+
+function search(targetNode, needle) {
+  if (needle.length < 3) {
+    // FIXME: We don't need to generate children if we haven't hit a search yet
+    //        however this is fairly fast and means we don't need to track searches.
+    generateChildren(targetNode);
+    return;
+  }
+
+  needle = needle.toLowerCase();
+
+  for (const child of Array.prototype.slice.call(targetNode.children)) {
+    child.remove();
+  }
+
+  const results = pathInTree.reduce((acc, x) => acc.children[x], tree);
+  generateSummary(results);
+
+  function collectSearchResults(result, allChildren = false, extraPath = "") {
+    return Object.entries(result).flatMap(([childName, child]) => {
+      const isLeaf = child.children === null;
+
+      let isSearchedFor = childName.toLowerCase().includes(needle);
+      let relativePath = extraPath + childName;
+      if (isLeaf) {
+        if (isSearchedFor) return [[relativePath, child]];
+        else return [];
+      }
+
+      const childrenResults = collectSearchResults(
+        child.children,
+        false,
+        relativePath + "/"
+      );
+      if (isSearchedFor) childrenResults.push([relativePath, child]);
+
+      return childrenResults;
+    });
+  }
+
+  const maxResultsShown = 500;
+  const foundResults = collectSearchResults(results.children).sort(
+    sortResultsByTypeAndName
+  );
+  foundResults
+    .filter((_, i) => i < maxResultsShown)
+    .forEach(([relativePath, child]) => {
+      const childNode = generateChildNode(
+        relativePath,
+        child,
+        `${pathInTree.join("/")}/${relativePath}`
+      );
+
+      if (child.children !== null) {
+        const extraPathParts = [
+          ...relativePath.split("/").filter((s) => s.length > 0),
+        ];
+        makeChildNavigable(childNode, extraPathParts, targetNode);
       }
+
+      targetNode.appendChild(childNode);
     });
+
+  if (foundResults.length > maxResultsShown) {
+    const extraNode = document.createElement("p");
+    extraNode.innerText = `Only displaying the first ${maxResultsShown} results! In total had ${foundResults.length} results.`;
```

Comment:
Hmm maybe it should be `In total _have found_ 999 results` or something like that?

Reply:
"Only showing the first X of Y total results"?

Reply:
Yeah that's better